src/main/resources/assets/js/app/inputtype/ui/text/HTMLAreaHelper.ts

```diff
@@ -97,4 +105,37 @@ export class HTMLAreaHelper {
 
         return processedContent;
     }
+
+    public static isSourceCodeEditable(): Q.Promise<boolean> {
+        return new IsAuthenticatedRequest().sendAndParse().then((loginResult: LoginResult) => {
+            if (loginResult.isContentExpert()) {
+                return Q(true);
+            }
+
+            return new ProjectGetRequest(ProjectContext.get().getProject()).sendAndParse().then((project: Project) => {
+                return Q(HTMLAreaHelper.isAllowedToEditSourceCode(loginResult, project));
+            });
+        }).catch((reason: any) => {
+            DefaultErrorHandler.handle(reason);
+            return Q(false);
+        });
+    }
+
+    private static isAllowedToEditSourceCode(loginResult: LoginResult, project: Project): boolean {
+        const userPrincipals: PrincipalKey[] = loginResult.getPrincipals();
+
+        const isExpert: boolean = project.getPermissions().getEditors().some((expert: PrincipalKey) => {
+            return userPrincipals.some((userPrincipal: PrincipalKey) => userPrincipal.equals(expert));
+        });
+
+        if (isExpert) {
+            return true;
+        }
+
+        const isOwner: boolean = project.getPermissions().getOwners().some((owner: PrincipalKey) => {
+            return userPrincipals.some((userPrincipal: PrincipalKey) => userPrincipal.equals(owner));
+        });
+
+        return isOwner;
+    }
```

Comment:
One pass of the `userPrincipals` array should be enough to find whether it's an editor or an owner:

```
const permissions = project.getPermissions();
const owners = permissions.getOwners();
const editors = permissions.getEditors();

return userPrincipals.some((userPrincipal: PrincipalKey) => 
  owners.indexOf(userPrincipal) > -1 || editors.indexOf(userPrincipal) > -1);
```